consensus/consensus_service.go

```diff
@@ -227,9 +228,62 @@ func (consensus *Consensus) VerifyHeaders(chain consensus_engine.ChainReader, he
 	return abort, results
 }
 
+// retrievePublicKeysFromLastBlock finds the public keys of last block's committee
+func retrievePublicKeysFromLastBlock(bc consensus_engine.ChainReader, header *types.Header) ([]*bls.PublicKey, error) {
+	parentHeader := bc.GetHeaderByHash(header.ParentHash)
+	if parentHeader == nil {
+		return nil, ctxerror.New("cannot find parent block header in DB",
+			"parentHash", header.ParentHash)
+	}
+	parentShardState, err := bc.ReadShardState(parentHeader.Epoch)
+	if err != nil {
+		return nil, ctxerror.New("cannot read shard state",
+			"epoch", parentHeader.Epoch,
+		).WithCause(err)
+	}
+	parentCommittee := parentShardState.FindCommitteeByID(parentHeader.ShardID)
+	if parentCommittee == nil {
+		return nil, ctxerror.New("cannot find shard in the shard state",
+			"parentBlockNumber", parentHeader.Number,
+			"shardID", parentHeader.ShardID,
+		)
+	}
+	var committerKeys []*bls.PublicKey
+	for _, member := range parentCommittee.NodeList {
+		committerKey := new(bls.PublicKey)
+		err := member.BlsPublicKey.ToLibBLSPublicKey(committerKey)
+		if err != nil {
+			return nil, ctxerror.New("cannot convert BLS public key",
+				"blsPublicKey", member.BlsPublicKey).WithCause(err)
+		}
+		committerKeys = append(committerKeys, committerKey)
+	}
+	return committerKeys, nil
+}
+
 // VerifySeal implements consensus.Engine, checking whether the given block satisfies
-// the PoW difficulty requirements.
+// the PoS difficulty requirements, i.e. >= 2f+1 valid signatures from the committee
 func (consensus *Consensus) VerifySeal(chain consensus_engine.ChainReader, header *types.Header) error {
+	publicKeys, err := retrievePublicKeysFromLastBlock(chain, header)
+	if err != nil {
+		return ctxerror.New("[VerifySeal] Cannot retrieve publickeys from last block").WithCause(err)
+	}
+	payload := append(header.LastCommitSignature[:], header.LastCommitBitmap...)
+	aggSig, mask, err := readSignatureBitmapByPublicKeys(payload, publicKeys)
+	if err != nil {
+		return ctxerror.New("[VerifySeal] Unable to deserialize the LastCommitSignature and LastCommitBitmap in Block Header").WithCause(err)
+	}
+	if count := utils.CountOneBits(mask.Bitmap); count < consensus.Quorum() {
```

Comment:
consensus.Quorum() is based on committee size of current block, not last block. At least, add a todo here if not fix it now.

Reply:
todo added